README
IWL-SWFUpload version 0.6.1
================

The IWL-SWFUpload package provides a special widget for the IWL library.
It provides the means for users to upload multiple files to a server,
using Adobe Flash 9. The package provides the base widget for browsing
and uploading files, as well as an additional helper widget, for visually
displaying the queued and uploaded files, as well as their status.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

In order to actually use the library, the javascript and css files will also
have to be installed in the server's document root. To do that, use the
'iwl-install' script, which is provided by IWL. It is usually located in '/usr/bin'

DEPENDENCIES

This module requires these other modules and libraries:

  IWL

COPYRIGHT AND LICENCE

Copyright (c) 2006-2007  Viktor Kojouharov. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See perldoc perlartistic.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.



Changes
Initial release: version 0.6.0 - 03.06.08



lib/IWL/SWFUpload.pm
#! /bin/false
# vim: set autoindent shiftwidth=4 tabstop=8:

package IWL::SWFUpload;

use strict;

use base 'IWL::Container';

use IWL::Button;
use IWL::JSON 'toJSON';
use IWL::String 'randomize';
use IWL::Config '%IWLConfig';

use vars qw($VERSION);

use Locale::TextDomain qw(org.bloka.iwl.swfupload);

$VERSION = '0.6.1';

my $init;

=head1 NAME

IWL::SWFUpload - a file upload widget using the flash swfupload library

=head1 INHERITANCE

L<IWL::Error> -> L<IWL::Object> -> L<IWL::Widget> -> L<IWL::Container> -> L<IWL::SWFUpload>

=head1 DESCRIPTION

The SWFUpload widget is a widget for upload files. It uses the flash swfupload library in order to achieve true multi-upload functionality. Visually, it's equivalent to a form's [Browse ...] button.

=head1 CONSTRUCTOR

IWL::SWFUpload->new ([B<%ARGS>])

Where B<%ARGS> is an optional hash parameter with with key-values.

=over 4

=item B<multiple>

If true, the invoked file selector can pick multiple files. Defaults to I<false>

=item B<autoUpload>

If true, file uploading starts as soon as the files are selected. An upload/stop button is never placed on the page.

=back

=head1 PROPERTIES 

The object has the following public properties

=over 4

=item B<browse>

A browse button, of class L<IWL::Button>. When pressed, it invokes the browser file selector.

=item B<upload>

An upload button, of class L<IWL::Button>. Before the upload button starts, it's action is to start it.

=item B<stop>

A stop upload button, of class L<IWL::Button>. After the upload start, this button can abort the process.

=back

=head1 SIGNALS

=over 4

=item B<load>

Fires when the widget is loaded.

=item B<file_dialog_start>

Fires immediately before the file selector is displayed. Due to the blocking nature of the file selector, the callback might not be executed until the selector is closed.

=item B<file_queue>

Fires when a file is queued. Receives the file object as a second parameter

=item B<file_queue_error>

Fires when an error has occurred while queueing a file. It received the file object, error code and error message as parameters, starting from the second parameters

=item B<file_dialog_complete>

Fires when the file selector is closed, and all the selected files have been processed. It received the number of selected files as a second parameter.

=item B<upload_start>

Fires just before the file is uploaded. It receives the file object as a second parameter.

=item B<upload_progress>

Fires periodically to indicate the upload process. It received the file object, the uploaded bytes, and the total byte count as parameters, starting from the second parameter.
A bug in the Linux Flash Player might prevent this signal from firing more than once.

=item B<upload_error>

Fires when an upload does not complete successfully, or by stopping/cancelling the upload. It receives the file object, error code and error message as parameters, starting from the second parameter.

=item B<upload_success>

Fires when the upload has been completed successfully. Receives the file object and any data returned from the server as parameters, starting from the second parameter.

=item B<upload_complete>

Fires at the end of the upload cycle. When fired, the upload will be fully completed, so that another may start. Receives the file object as a second parameter.

=back

=cut

sub new {
    my ($proto, %args) = @_;
    my $class = ref($proto) || $proto;
    my $self  = $class->SUPER::new;

    $init->($self, %args);

    return $self;
}

=head1 METHODS

=over 4

=item B<setMultiple> (B<BOOL>)

Sets whether the widget's file selector can pick multiple files

Parameters: B<BOOL> - if true, the selector can pick multiple files.

=cut

sub setMultiple {
    my ($self, $bool) = @_;

    $self->{_options}{multiple} = !(!$bool);
    return $self;
}

=item B<isMultiple>

Returns true if the file selector can pick multiple files

=cut

sub isMultiple {
    return shift->{_options}{multiple};
}

=item B<setUploadURL> (B<URL>)

Sets the upload script to call

Parameters: B<URL> - the url to call on file upload

=cut

sub setUploadURL {
    my ($self, $url) = @_;

    $self->{__SWFOptions}{upload_url} = $url;

    return $self;
}

=item B<getUploadURL>

Returns the url of the upload script

=cut

sub getUploadURL {
    return shift->{__SWFOptions}{upload_url};
}

=item B<setPostParams> (B<%PARAMS>)

Sets additional post parameters to pass to the script

Parameters: B<%PARAMS> - a hash of parameters

=cut

sub setPostParams {
    my ($self, %params) = @_;

    $self->{__SWFOptions}{post_params} = {%{$self->{__SWFOptions}{post_params} || {}}, %params};

    return $self;
}

=item B<getPostParams>

Returns the post parameters

=cut

sub getPostParams {
    return %{shift->{__SWFOptions}{post_params}};
}

=item B<setFileTypes> (B<TYPES>, B<DESCRIPTION>)

Sets the allowed file types for upload.

Parameters: B<TYPES> - the allowed types. A semicolor separater string of glob patterns. See glob(3). B<DESCRIPTION> - a description, which is displayed in the file selector

=cut

sub setFileTypes {
    my ($self, $types, $description) = @_;

    $self->{__SWFOptions}{file_types} = $types;
    $self->{__SWFOptions}{file_types_description} = $description;

    return $self;
}

=item B<getFileTypes>

Returns the allowed file types and the description.

=cut

sub getFileTypes {
    my $self = shift;
    return $self->{__SWFOptions}{file_types}, $self->{__SWFOptions}{file_types_description};
}

=item B<setFileSizeLimit> (B<LIMIT>)

Sets the maximum size of the files to be uploaded.

Parameters: B<LIMIT> - the size limit, in I<KB>, unless a unit is specified. Examples:
  147483648 B, 2097152, 2097152KB, 2048 MB, 2 GB

=cut

sub setFileSizeLimit {
    my ($self, $limit) = @_;

    $self->{__SWFOptions}{file_size_limit} = $limit;

    return $self;
}

=item B<getFileSizeLimit>

Returns the maximum size of the files to be uploaded

=cut

sub getFileSizeLimit {
    return shift->{__SWFOptions}{file_size_limit};
}

=item B<setFileUploadLimit> (B<LIMIT>)

Sets the maximum number of files that can be uploaded

Parameters: B<LIMIT> - the number of files to be uploaded, or I<0> for unlimited uploading.

=cut

sub setFileUploadLimit {
    my ($self, $limit) = @_;

    $self->{__SWFOptions}{file_upload_limit} = $limit;

    return $self;
}

=item B<getFileUploadLimit>

Returns the maximum files that can be uploaded

=cut

sub getFileUploadLimit {
    return shift->{__SWFOptions}{file_upload_limit};
}

=item B<setFileQueueLimit> (B<LIMIT>)

Sets the maximum number of files that can be queued

Parameters: B<LIMIT> - the limit number. I<0> for an unlimited queue

=cut

sub setFileQueueLimit {
    my ($self, $limit) = @_;

    $self->{__SWFOptions}{file_queue_limit} = $limit;

    return $self;
}

=item B<getFileQueueLimit>

Returns the maximum number of files that can be queued

=cut

sub getFileQueueLimit {
    return shift->{__SWFOptions}{file_queue_limit};
}

=item B<registerPlugin> (B<PLUGIN>)

Registers an SWFUpload plugin. The following plugins are currently available, and documented within their respective JavaScript files:

=over 8

=item B<cookie>

If this plugin is registered, it will automatically add the cookie string to a POST parameter, named I<SWFUploadCookie>.

=back

Parameters: B<PLUGIN> - the name (or a list of names) of the plugin to register.

=cut

sub registerPlugin {
    my ($self, @plugin) = @_;
    my @plugins = map {'plugins/swfupload/' . $_ . '.js'} @plugin;

    push @{$self->{__plugins}}, @plugins;
    $self->requiredJs(@plugins);
    return $self;
}

# Protected
#
sub _realize {
    my $self  = shift;
    my $id    = $self->getId;

    $self->SUPER::_realize;
    return $self->_pushFatalError(__"SWFUpload: Upload URL not set!") unless $self->getUploadURL;

    my $messages = {
        progress => {
            queue    => __"Queued",
            complete => __"Completed",
            progress => "#{percent}",
            error    => __"Error",
        },
        uploadErrors => {
            -200 => __"HTTP error",
            -210 => __"Missing upload URL",
            -220 => __"IO error",
            -230 => __"Security error",
            -240 => __"Upload limit exceeded",
            -250 => __"Upload failed",
            -260 => __"Specified file ID not found",
            -270 => __"File validation failed",
            -280 => __"Upload cancelled",
            -290 => __"Upload stopped",
        },
        queueErrors => {
            -100 => __"Queue limit exceeded",
            -110 => __"File size limit exceeded",
            -120 => __"File has zero byte size",
            -130 => __"Invalid filetype",
        },
        buttonLabels => {
            start => __"Upload",
            stop  => __"Stop uploading",
        },
        flashErrors => {
            missingPlugin     => __"In order to upload files, you will need to upgrade to the latest version of the Adobe Flash Player.",
            missingPluginLink => __"Get the Adobe Flash Player",
        }
    };
    my $swfoptions = toJSON($self->{__SWFOptions});
    my $options = toJSON($self->{_options});
    $messages = toJSON($messages);
    $self->_appendInitScript("IWL.SWFUpload.create('$id', $swfoptions, $options, $messages)");

    $self->{upload}->remove if $self->{_options}{autoUpload};
}

sub _setupDefaultClass {
    my $self = shift;
    $self->prependClass($self->{_defaultClass});
    $self->{browse}->prependClass($self->{_defaultClass} . '_browse');
    $self->{upload}->prependClass($self->{_defaultClass} . '_upload');
}

# Internal
#
$init = sub {
    my ($self, %args) = @_;
    my $browse = IWL::Button->new;
    my $upload = IWL::Button->new;
    
    $self->{_options} = {};
    $self->{_options}{multiple}     = $args{multiple} ? 1 : 0;
    $self->{_options}{autoUpload}   = $args{autoUpload} ? 1 : 0;
    delete @args{qw(multiple autoUpload)};

    $browse->setLabel(__"Browse ...")->setDisabled(1);
    $upload->setLabel(__"Upload")->setDisabled(1);
    $self->{browse} = $browse;
    $self->{upload} = $upload;
    $self->appendChild($browse, $upload);
    $self->{_defaultClass} = 'swfupload';
    $args{id} ||= randomize($self->{_defaultClass});

    $self->{__SWFOptions} = {flash_url => $IWLConfig{JS_DIR} . '/dist/swfupload_f9.swf'};
    $self->{__plugins} = [];
    $self->requiredJs('base.js', 'dist/swfupload.js', 'dist/AC_OETags.js', 'swfupload.js');
    $self->_constructorArguments(%args);
    $self->{_customSignals} = {
        load => [],
        flash_not_found => [],
        file_dialog_start => [],
        file_queue => [],
        file_queue_error => [],
        file_dialog_complete => [],
        upload_start => [],
        upload_progress => [],
        upload_error => [],
        upload_success => [],
        upload_complete => [],
    };

    return $self;
};

1;

=head1 AUTHOR

  Viktor Kojouharov

=head1 Website

L<http://code.google.com/p/iwl>

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2006-2007  Viktor Kojouharov. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See perldoc perlartistic.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut


lib/IWL/SWFUpload/Queue.pm
#! /bin/false
# vim: set autoindent shiftwidth=4 tabstop=8:

package IWL::SWFUpload::Queue;

use strict;

use base 'IWL::Tree';

use IWL::Tree::Row;
use IWL::ProgressBar;

use IWL::JSON 'toJSON';
use IWL::String 'randomize';

use Locale::TextDomain qw(org.bloka.iwl.swfupload);

my $init;

=head1 NAME

IWL::SWFUpload::Queue - a file queue for the L<IWL::SWFUpload> widget

=head1 INHERITANCE

L<IWL::Error> -> L<IWL::Object> -> L<IWL::Widget> -> L<IWL::Tree> -> L<IWL::SWFUpload::Queue>

=head1 DESCRIPTION

The Queue widget stores the queued files for an L<IWL::SWFUpload> widget. It also allows for removing of individual files from the queue, as well as progress indication for each file.

=head1 CONSTRUCTOR

IWL::SWFUpload->new ([B<%ARGS>])

Where B<%ARGS> is an optional hash parameter with with key-values.

=over 4

=item B<order>

Describes the column order, as well as what colums are visible. The sytax is an array reference. Possible values are: 

=over 8

=item B<name>

The file name

=item B<status>

The status of the upload

=item B<start>

An image, which will start the upload of the file when clicked

=item B<stop>

An image, which will stop the upload of the file when clicked

=item B<remove>

An image, which will remove the file when clicked

=item B<{name =E<gt> name, callback =E<gt> callback, title =E<gt> title}>

If the element is a hash reference, it will be used in the following way: If a I<title> key is present, its value will be shown in the header. A I<callback> key, with a valid javascript function name must be present. The callback will be executed each time a file is queued. The table cell, file and upload objects will be passed to the callback upon execution. A I<name> key, specifying the name of the column, must also be present.

=back

Default value: I<[name, status, remove]>

=item B<showHeader>

If true, will show the queue header

=back

=head1 SIGNALS

=over 4

=item B<load>

Fires when the widget is loaded.

=back

=cut

sub new {
    my ($proto, %args) = @_;
    my $class = ref($proto) || $proto;
    my $self  = $class->SUPER::new(list => 1);

    $init->($self, %args);

    return $self;
}

=head1 METHODS

=over 4

=item B<bindToUpload> (B<UPLOAD>)

Binds the queue to the given upload widget.

Parameters: B<UPLOAD> - an L<IWL::SWFUpload> widget, or a string, which represents the L<IWL::SWFUpload> widget's ID.

=cut

sub bindToUpload {
    my ($self, $upload) = @_;

    return unless $upload
      && !ref $upload 
      || UNIVERSAL::isa($upload, 'IWL::SWFUpload');
    $self->{__upload} = $upload;

    return $self;
}

# Protected
#
sub _realize {
    my $self = shift;
    my $id   = $self->getId;
    my $progress;

    return $self->_pushFatalError(__"Queue: Missing SWFUpload widget") unless $self->{__upload};
    $self->SUPER::_realize;

    foreach my $column (@{$self->{__queueOptions}{order}}) {
        if (ref $column eq 'HASH' && $column->{name}) {
            $self->{__header}->appendTextHeaderCell($column->{title} || '');
        } elsif ($column eq 'name') {
            $self->{__header}->appendTextHeaderCell(__"Name");
        } elsif ($column eq 'status') {
            $self->{__header}->appendTextHeaderCell(__"Status");
            $progress = 1;
        } elsif (grep {$_ eq $column} qw(start stop remove)) {
            $self->{__header}->appendHeaderCell;
        } else {
            $column = '';
        }
    }
    if ($progress) {
        $self->{__progressBar}->setId($id . '_progress')->setStyle(display => 'none');
        unshift @{$self->{_tailObjects}}, $self->{__progressBar};
    }
    $self->{__queueOptions}{order} = [grep { !(!$_) } @{$self->{__queueOptions}{order}}];
    my $options = toJSON($self->{__queueOptions});
    my $uploadId = ref $self->{__upload}
      ? $self->{__upload}->getId
      : $self->{__upload};
    $self->_appendInitScript("IWL.SWFUpload.Queue.create('$id', '$uploadId', $options)");
    $self->{__header}->setStyle(display => 'none') unless $self->{__queueOptions}{showHeader};
}

# Internal
#
$init = sub {
    my ($self, %args) = @_;
    my $header = IWL::Tree::Row->new;
    my $progress = IWL::ProgressBar->new;
    
    $self->{__queueOptions} = {order => [qw(name status remove)]};
    $self->{__queueOptions}{order} = $args{order} if defined $args{order};
    $self->{__queueOptions}{showHeader} = $args{showHeader} ? 1 : 0;
    delete @args{qw(order showHeader)};

    $self->appendClass('swfupload_queue');
    $args{id} ||= randomize($self->{_defaultClass});

    $self->{__progressBar} = $progress;
    $self->{__header} = $header;
    $self->appendHeader($header);
    $self->requiredJs('base.js', 'dist/swfupload.js', 'swfupload.js');
    $self->_constructorArguments(%args);

    $self->{_customSignals} = {
        load => [],
    };
    return $self;
};

1;

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2006-2007  Viktor Kojouharov. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See perldoc perlartistic.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut


inc/Module/Install/Share.pm
package Module::Install::Share;

use strict;
use Module::Install::Base;

use vars qw{$VERSION $ISCORE @ISA};
BEGIN {
	$VERSION = '0.67';
	$ISCORE  = 1;
	@ISA     = qw{Module::Install::Base};
}

sub install_share {
	my ($self, $dir) = @_;

	if ( ! defined $dir ) {
		die "Cannot find the 'share' directory" unless -d 'share';
		$dir = 'share';
	}

	$self->postamble(<<"END_MAKEFILE");
INST_AUTODIR2   = \$(INST_LIB)/auto/IWL
config ::
\t\$(NOECHO) \$(MOD_INSTALL) \\
\t\t"$dir" \$(INST_AUTODIR2)

END_MAKEFILE

	# The above appears to behave incorrectly when used with old versions
	# of ExtUtils::Install (known-bad on RHEL 3, with 5.8.0)
	# So when we need to install a share directory, make sure we add a
	# dependency on a moderately new version of ExtUtils::MakeMaker.
	$self->build_requires( 'ExtUtils::MakeMaker' => '6.11' );
}

1;

__END__

=pod

=head1 NAME

Module::Install::Share - Install non-code files for use during run-time

=head1 SYNOPSIS

    # Put everything inside ./share/ into the distribution 'auto' path
    install_share 'share';

    # Same thing as above using the default directory name
    install_share;

=head1 DESCRIPTION

As well as Perl modules and Perl binary applications, some distributions
need to install read-only data files to a location on the file system
for use at run-time.

XML Schemas, L<YAML> data files, and L<SQLite> databases are examples of
the sort of things distributions might typically need to have available
after installation.

C<Module::Install::Share> is a L<Module::Install> extension that provides
commands to allow these files to be installed to the applicable location
on disk.

To locate the files after installation so they can be used inside your
module, see this extension's companion module L<File::ShareDir>.

=head1 TO DO

Currently C<install_share> installs not only the files you want, but
if called by the author will also copy F<.svn> and other source-control
directories, and other junk.

Enhance this to copy only files under F<share> that are in the
F<MANIFEST>, or possibly those not in F<MANIFEST.SKIP>.

=head1 AUTHORS

Audrey Tang E<lt>autrijus@autrijus.orgE<gt>

Adam Kennedy E<lt>cpan@ali.asE<gt>

=head1 SEE ALSO

L<Module::Install>, L<File::ShareDir>

=head1 COPYRIGHT

Copyright 2006 Audrey Tang, Adam Kennedy. All rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

See L<http://www.perl.com/perl/misc/Artistic.html>

=cut


share/jscript/swfupload.js
// vim: set autoindent shiftwidth=4 tabstop=8:
/**
 * @class IWL.SWFUpload is a class for creating flash upload widgets
 * @extends IWL.Widget
 * */
IWL.SWFUpload = Object.extend(Object.extend({}, IWL.Widget), (function () {
    var effect_duration = 0.06;

    function completeHandler() {
        var stats = this.control.getStats();
        if (stats.files_queued) {
            if (!this.options.autoUpload && !this.upload._uploadStarted) return;
            this.control.startUpload.bind(this.control).defer()
        } else {
            if (this.options.autoUpload) return;
            this.upload.setDisabled(1);
            if (!this.upload._uploadStarted) return;
            buttonToggle.call(this, this.upload._uploadStarted = false);
        }
    }

    function queueHandler() {
        var stats = this.control.getStats();
        if (stats.files_queued)
            this.upload.setDisabled(0);
        else
            this.upload.setDisabled(1);
    }

    function uploadStartHandler(event, file) {
        this.upload.setDisabled(0);
    }

    function uploadErrorHandler(event, file, code) {
        var stats = this.control.getStats();
        if (code != SWFUpload.UPLOAD_ERROR.FILE_CANCELLED) return;
        if (stats.files_queued) return;

        buttonToggle.call(this, false);
        this.upload.setDisabled(1);
    }

    function uploadProgressHandler(event, file, complete, total) {
        if (complete/total == 1)
            this.upload.setDisabled(1);
    }

    function buttonToggle(start) {
        this.upload.setLabel(IWL.SWFUpload.messages.buttonLabels[start ? 'stop' : 'start']);
    }

    return {

        _init: function (id, swfoptions) {
            this.options = Object.extend({
                multiple: false,
                autoUpload: false
            }, arguments[2] || {});
            Object.extend(IWL.SWFUpload, {messages: arguments[3]});

            var className = $A(this.classNames()).first();

            // This is fixed in SWFUpload 2.1.0, and should be removed once ported to this version
            if (swfoptions.post_params && Object.isObject(swfoptions.post_params))
                for (var i in swfoptions.post_params)
                    swfoptions.post_params[i] = swfoptions.post_params[i].toString();

            this.control = new SWFUpload(Object.extend(swfoptions, {
                swfupload_loaded_handler: this.emitSignal.bind(this, 'iwl:load'),
                file_dialog_start_handler: this.emitSignal.bind(this, 'iwl:file_dialog_start'),
                file_queued_handler: this.emitSignal.bind(this, 'iwl:file_queue'),
                file_queue_error_handler: this.emitSignal.bind(this, 'iwl:file_queue_error'),
                file_dialog_complete_handler: this.emitSignal.bind(this, 'iwl:file_dialog_complete'),
                upload_start_handler: 
                    function(file) {
                        this.emitSignal('iwl:upload_start', file);
                        return true;
                    }.bind(this),
                upload_progress_handler: this.emitSignal.bind(this, 'iwl:upload_progress'),
                upload_error_handler: this.emitSignal.bind(this, 'iwl:upload_error'),
                upload_success_handler: this.emitSignal.bind(this, 'iwl:upload_success'),
                upload_complete_handler: this.emitSignal.bind(this, 'iwl:upload_complete')
            }));
            this.browse = this.select('.' + className + '_browse')[0];

            if (this.options.autoUpload) {
                this.signalConnect('iwl:file_dialog_complete', function() {this.control.startUpload()}.bind(this));
            } else {
                this.upload = this.select('.' + className + '_upload')[0];
                this.upload._uploadStarted = false;

                this.upload.signalConnect('click', function() {
                    buttonToggle.call(this, this.upload._uploadStarted = !this.upload._uploadStarted);
                    this.upload._uploadStarted
                      ? this.control.startUpload()
                      : this.control.stopUpload();
                }.bind(this));
                this.signalConnect('iwl:file_queue', queueHandler.bind(this));
                this.signalConnect('iwl:upload_start', uploadStartHandler.bind(this));
                this.signalConnect('iwl:upload_error', uploadErrorHandler.bind(this));
                this.signalConnect('iwl:upload_progress', uploadProgressHandler.bind(this));
            }

            this.browse.signalConnect('click', function() {
                this.options.multiple ? this.control.selectFiles() : this.control.selectFile();
            }.bind(this));
            this.signalConnect('iwl:upload_complete', completeHandler.bind(this));

            if (DetectFlashVer(9, 0, 0))
                document.observe('dom:loaded', function() {
                    this.browse.setDisabled(false);
                }.bind(this));
            else {
                document.observe('dom:loaded', function() {
                    this.emitSignal(this, 'iwl:flash_not_found')
                    var info = new Element('span', {className: 'swfupload_missing_flash_plugin'}).update(IWL.SWFUpload.messages.flashErrors.missingPlugin);
                    info.appendChild(new Element('br'));
                    info.appendChild(new Element('a', {
                        href: 'http://www.adobe.com/go/getflashplayer',
                        target: 'ADOBEFLASH'
                    }).update(IWL.SWFUpload.messages.flashErrors.missingPluginLink));
                    
                    this.insert({after: info});
                }.bind(this));
            }
        }
    }
})());

/**
 * @class IWL.SWFUpload.Queue is a class for creating a queue for the flash upload widgets
 * @extends IWL.Widget
 * */
IWL.SWFUpload.Queue = Object.extend(Object.extend({}, IWL.Widget), (function () {
    var disabled_opacity = 0.2;

    function fileQueueHandler(event, file) {
        var id = this.id + '_' + file.id;
        var names = {id: id};
        var className = $A(this.classNames()).first();
        var progress = false, images = [], custom = [];
        this.options.order.each(function(column, i) {
            if (column == 'status') progress = true;
            if (['start', 'stop', 'remove'].include(column))
                images.push(column);
            if (Object.isObject(column)) {
                custom.push(column);
                names['name' + i] = column.name;
            } else {
                names['name' + i] = column;
            }
        });
        var html = this.template.evaluate(names);
        this.appendRow(this.body, html);
        var row = $(id);
        var cell;

        this.files[file.id] = {row: row};
        if (progress) {
            cell = row.select('.' + className + '_status')[0];
            progress = this.progress.cloneNode(true);
            IWL.ProgressBar.create(progress);
            cell.appendChild(progress);
            progress.appear({duration: 0.3});
            progress.id = id + '_progress';
            progress.setText(IWL.SWFUpload.messages.progress.queue);
            row.progress = progress;
        }
        images.each(function(column) {
            cell = row.select('.' + className + '_' + column)[0];
            var image = new Element('img', {
                src: IWL.Config.IMAGE_DIR + '/queue/' + column + '.' + IWL.Config.ICON_EXT,
                alt: column,
                id: id + '_' + column
            });
            cell.appendChild(image);
            if (column == 'remove') {
                row.removeCell = cell;
                var removeHandler = (function() {
                    this.upload.control.cancelUpload(file.id);
                }).bind(this);
                cell.handler = removeHandler;
                cellToggle(cell, true);
            } else if (column == 'stop') {
                row.stopCell = cell;
                var stopHandler = (function() {
                    cellToggle(row.stopCell, false);
                    if (row.startCell)
                        cellToggle(row.startCell, true);
                    this.upload.control.stopUpload();
                }).bind(this);
                cell.handler = stopHandler;
                cellToggle(cell, false);
            } else if (column == 'start') {
                row.startCell = cell;
                var startHandler = (function() {
                    cellToggle(row.startCell, false);
                    if (row.stopCell)
                        cellToggle(row.stopCell, true);
                    this.upload.control.startUpload(file.id);
                }).bind(this);
                cell.handler = startHandler;
                cellToggle(cell, true);
            }
        }.bind(this));
        custom.each(function(column) {
            cell = row.select('.' + className + '_' + column.name)[0];
            var callback = codePointer(column.callback);
            if (Object.isFunction(callback))
                callback(cell, file, this.upload);
        }.bind(this));

        cell = row.select('.' + className + '_name')[0];
        if (cell) cell.update(file.name);
        return this;
    }

    function uploadStartHandler(event, file) {
        var row = this.files[file.id].row;
        if (row.progress)
            row.progress.setText(IWL.SWFUpload.messages.progress.progress).setValue(0);
        if (row.startCell)
            cellToggle(row.startCell, false);
        if (row.stopCell)
            cellToggle(row.stopCell, true);
    }

    function uploadProgressHandler(event, file, complete, total) {
        var row = this.files[file.id].row;
        if (row.progress)
            row.progress.setValue(complete/total);
    }

    function uploadErrorHandler(event, file, code) {
        var row = this.files[file.id].row;
        if (!row || !row.parentNode) return;
        if (row.progress)
            row.progress.setText(
                ([-280, -290].include(code)
                  ? ''
                  : (IWL.SWFUpload.messages.progress.error + ': ')) +
                IWL.SWFUpload.messages.uploadErrors[code]
            );
        if (code == SWFUpload.UPLOAD_ERROR.UPLOAD_STOPPED) {
            if (row.startCell)
                cellToggle(row.startCell, true);
            if (row.stopCell)
                cellToggle(row.stopCell, false);
            if (row.progress)
                (function() {
                    row.progress.setText(IWL.SWFUpload.messages.progress.queue);
                }).delay(3);
        } else if (code == SWFUpload.UPLOAD_ERROR.FILE_CANCELLED) {
            row.fade.bind(row, {duration: 1, afterFinish: row.remove.bind(row)}).delay(3);
        }
    }

    function uploadSuccessHandler(event, file) {
        var row = this.files[file.id].row;
        if (row.progress)
            row.progress.setText(IWL.SWFUpload.messages.progress.complete);
        if (row.stopCell)
            cellToggle(row.stopCell, false);
        if (row.removeCell)
            cellToggle(row.removeCell, false);
    }

    function cellToggle(cell, on) {
        if (on) {
            cell.observe('click', cell.handler);
            cell.setOpacity(1);
        } else {
            cell.stopObserving('click', cell.handler);
            cell.setOpacity(disabled_opacity);
        }
    }

    function codePointer(string) {
        if (!Object.isString(string)) return;
        return string.split('.').inject(window, function(parent, child) {
            if (!parent || !parent[child]) throw $break;
            return parent[child];
        });
    }

    return {
        /**
         * Toggles the enable upload state per queue file.
         * @param {Object} options The following options are recognized:
         *        fileId: The the file id, corresponding to the file to toggle. If none are specified, all files are toggled.
         *        on: if true, upload will be allowed. Otherwise, upload will be disabled.
         *
         * @returns The object
         * */
        toggleFileUploadState: function() {
            var options = Object.extend({on: false}, arguments[0]), rows = [];
            if (options.fileId) rows.push(this.files[options.fileId].row);
            else rows = $H(this.files).values().pluck('row');
            rows.each(function(row) {
                if (!row.parentNode || !row.startCell) return;
                cellToggle(row.startCell, options.on);
            });
        },

        _preInit: function(id, upload) {
            if (!$(upload) && !document.loaded) {
                document.observe('dom:loaded', this.create.bind(this, id, upload, arguments[2]));
                return false;
            }
            return true;
        },

        _init: function (id, upload) {
            this.options = Object.extend({
                order: ['name', 'status', 'remove']
            }, arguments[2] || {});
            if (Object.isString(this.options.order))
                this.options.order = this.options.order.evalJSON(1);

            this.progress = $(id + '_progress');

            var className = $A(this.classNames()).first();
            var template = '<tr id="#{id}">';
            this.options.order.length.times(function(i) {
                template += '<td class="' + className + '_#{name' + i + '}"></td>';
            });
            template += '</tr>';
            this.template = new Template(template);

            this.files = {};
            this.upload = upload = $(upload);
            upload.signalConnect('iwl:file_queue', fileQueueHandler.bind(this));
            upload.signalConnect('iwl:upload_start', uploadStartHandler.bind(this));
            upload.signalConnect('iwl:upload_progress', uploadProgressHandler.bind(this));
            upload.signalConnect('iwl:upload_error', uploadErrorHandler.bind(this));
            upload.signalConnect('iwl:upload_success', uploadSuccessHandler.bind(this));

            this.emitSignal('iwl:load');
        }
    }
})());
